iroh-net/src/relay/codec.rs

```diff
@@ -264,8 +264,67 @@ impl Frame {
         }
     }
 
+    fn from_ws_vec(vec: Vec<u8>) -> anyhow::Result<Self> {
+        if vec.is_empty() {
+            bail!("error parsing relay::codec::Frame: too few bytes (0)");
+        }
+        let bytes = Bytes::from(vec);
+        let typ = FrameType::from(bytes[0]);
+        let frame = Self::from_bytes(typ, bytes.slice(1..))?;
+        Ok(frame)
+    }
+
+    fn into_ws_vec(self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        bytes.put_u8(self.typ().into());
+        self.write_to(&mut bytes);
+        bytes
+    }
+
+    /// Serializes this to bytes and wraps it in a websocket binary message.
+    pub(crate) fn into_ws_message(self) -> std::io::Result<tungstenite::Message> {
+        Ok(tungstenite::Message::binary(self.into_ws_vec()))
+    }
+
+    /// Serializes this to bytes and wraps it in a websocket binary message for use in wasm.
+    pub(crate) fn into_wasm_ws_message(self) -> std::io::Result<tokio_tungstenite_wasm::Message> {
+        Ok(tokio_tungstenite_wasm::Message::binary(self.into_ws_vec()))
+    }
+
+    /// Unwraps any binary messages received via websockets and parses them into `Frame`s.
+    ///
+    /// Ignores any non-binary websocket messages with a warning.
+    pub(crate) fn from_ws_message(
+        msg: tungstenite::Result<tungstenite::Message>,
+    ) -> Option<anyhow::Result<Self>> {
+        match msg {
+            Ok(tungstenite::Message::Binary(vec)) => Some(Self::from_ws_vec(vec)),
+            Ok(msg) => {
+                tracing::warn!(?msg, "Got websocket message of unsupported type, skipping.");
+                None
+            }
+            Err(e) => Some(Err(e.into())),
+        }
+    }
+
+    /// Unwraps any binary messages received via websockets (in wasm) and parses them into `Frame`s.
+    ///
+    /// Ignores any non-binary websocket messages with a warning.
+    pub(crate) fn from_wasm_ws_message(
+        msg: tokio_tungstenite_wasm::Result<tokio_tungstenite_wasm::Message>,
+    ) -> Option<anyhow::Result<Self>> {
+        match msg {
+            Ok(tokio_tungstenite_wasm::Message::Binary(vec)) => Some(Self::from_ws_vec(vec)),
+            Ok(msg) => {
+                tracing::warn!(?msg, "Got websocket message of unsupported type, skipping.");
+                None
+            }
+            Err(e) => Some(Err(e.into())),
+        }
+    }
+
     /// Writes it self to the given buffer.
-    fn write_to(&self, dst: &mut BytesMut) {
+    fn write_to(&self, mut dst: impl BufMut) {
```

Comment:
While this work I think it's more correct to use
```suggestion
    fn write_to(&self, dst: &mut impl BufMut) {
```

Reply:
Hmm. I've thought about it a bunch and am on the fence. But I agree that it's slightly less odd to see `&mut` here, and it's a little better because you don't have to care that much about giving away ownership on the inside of the function accidentally :)